File: service/game_service.go
package service

import (
	"steamtrender.com/api/repository"
	"steamtrender.com/api/schema"
	"steamtrender.com/api/utils"
)

type IGameService interface {
	IService
	AnalyzeCompetitors(float64, int, int, int, []uint, []uint) (schema.CompetitorsData, error)
	AnalyzeTags(float64, int, int, int, []uint) ([]schema.CompetitorsData, error)
}

type GameService struct {
	Repo repository.IGameRepository
}

func NewGameService(repo *repository.GameRepository) *GameService {
	return &GameService{Repo: repo}
}

func (service *GameService) AnalyzeCompetitors(reviwsCoeff float64, minReviews int, minYear int, maxYear int, whitelistTagIDs, blacklistTagIDs []uint) (schema.CompetitorsData, error) {
	games, err := service.Repo.ReadGames(minReviews, minYear, maxYear, whitelistTagIDs, blacklistTagIDs)
	if err != nil {
		return schema.CompetitorsData{}, err
	}

	data := service.AnalyzeGames(games, reviwsCoeff)

	return data, nil
}

func (service *GameService) AnalyzeTags(reviwsCoeff float64, minYear, maxYear, reviews int, tags []uint) ([]schema.CompetitorsData, error) {
	var results []schema.CompetitorsData
	// fix it ^^^
	for _, tag := range tags {
		fakeTags := []uint{tag}
		games, err := service.Repo.ReadGames(reviews, minYear, maxYear, fakeTags, nil)
		if err != nil {
			return nil, err
		}

		data := service.AnalyzeGames(games, reviwsCoeff)
		results = append(results, data)
	}

	return results, nil
}

func (gs *GameService) AnalyzeGames(games []schema.Game, reviewsCoeff float64) schema.CompetitorsData {
	data := schema.CompetitorsData{
		Games:      games,
		TotalGames: int64(len(games)),
	}

	if len(games) > 0 {
		var reviews []float64
		var owners []float64
		var prices []float64
		var revenues []float64

		for _, game := range games {
			reviews = append(reviews, float64(game.Reviews))
			owners = append(owners, reviews[len(reviews)-1]*reviewsCoeff)
			prices = append(prices, float64(game.Price))
			revenues = append(revenues, prices[len(prices)-1]*owners[len(owners)-1]*utils.RevenueCoeff())
		}

		data.MedianReviews = utils.CalculateMedian(reviews)
		data.MedianOwners = utils.CalculateMedian(owners)
		data.MedianPrice = utils.CalculateMedian(prices)
		data.MedianRevenue = utils.CalculateMedian(revenues)
	}

	return data
}

var _ IGameService = &GameService{}

File: services/game_service.py
from datetime import date
from typing import List

import numpy as np
from sqlalchemy import asc, desc, func
from sqlalchemy.future import select
from sqlalchemy.orm import selectinload

from models.game import Game
from models.tag import Tag, GameTagAssociation
from schema.games_overview import GamesOverview
from schema.revenue import Revenue
from schema.tag import ExtendedTag
from utils.constants import RevenueCoeff
from utils.validate_range import validate_range


class GameService:
    @staticmethod
    async def read_all_tags(session) -> List[Tag]:
        query = select(Tag).order_by(asc(Tag.title))
        result = await session.execute(query)
        tags = result.scalars().all()
        return tags

    @staticmethod
    async def read_tag_by_id(session, tag_id) -> Tag:
        query = select(Tag).filter(Tag.id == tag_id)
        result = await session.execute(query)
        tag = result.scalars().first()
        return tag

    @staticmethod
    async def read_games(
        session,
        min_reviews: int = 0,
        max_reviews: int = None,
        min_price: float = 0,
        max_price: float = None,
        min_date: date = date(2020, 1, 1),
        max_date: date = date(2024, 12, 31),
        whitelist_tag_ids: list = None,
        blacklist_tag_ids: list = None,
        tag_threshold: int = 5,
    ) -> List[Game]:
        min_reviews, max_reviews = validate_range(min_reviews, max_reviews)
        min_price, max_price = validate_range(min_price, max_price)
        min_date, max_date = validate_range(min_date, max_date)

        query = select(Game).where(Game.reviews >= min_reviews)
        if max_reviews:
            query = query.where(Game.reviews <= max_reviews)

        query = query.where(Game.price >= min_price)
        if max_price:
            query = query.where(Game.price <= max_price)

        query = query.where(Game.release_date.between(min_date, max_date))

        if whitelist_tag_ids:
            for tag_id in whitelist_tag_ids:
                query = query.filter(
                    Game.tag_associations.any(
                        (GameTagAssociation.tag_id == tag_id) &
                        (GameTagAssociation.tag_number <= tag_threshold)
                    )
                )

        if blacklist_tag_ids:
            for tag_id in blacklist_tag_ids:
                query = query.filter(
                    ~Game.tag_associations.any(
                        (GameTagAssociation.tag_id == tag_id) &
                        (GameTagAssociation.tag_number <= tag_threshold)
                    )
                )

        query = query.options(
            selectinload(Game.tag_associations).selectinload(GameTagAssociation.tag)
        )

        query = query.order_by(desc(Game.reviews))
        result = await session.execute(query)
        games = result.scalars().all()

        print(len(games))

        return games

    @staticmethod
    async def analyze_games(
        games: List[Game],
        reviews_coeff: float,
        revenue_agg: List[float],
        update_games: bool = False,
    ) -> GamesOverview:
        data = GamesOverview(total_games=len(games))
        if not games:
            return data

        reviews = np.array([game.reviews for game in games])
        prices = np.array([game.price for game in games])

        owners = reviews * reviews_coeff
        revenues = np.round(owners * prices * RevenueCoeff)

        if update_games:
            for i, game in enumerate(games):
                game.owners = owners[i]
                game.revenue = revenues[i]

        data.median_reviews = int(np.median(reviews))
        data.median_owners = int(np.median(owners))
        data.median_price = round(float(np.median(prices)), 2)
        data.revenue_total = int(np.sum(revenues))

        for agg in revenue_agg:
            rev_agg = Revenue(agg=agg, value=int(np.quantile(revenues, agg)))
            data.revenue.append(rev_agg)

        return data

    @staticmethod
    async def get_top_tags(
        session, games: List[Game], skip_tag_ids: List[int], limit: int = 3
    ) -> List[ExtendedTag]:
        game_ids = [game.id for game in games]

        stmt = (
            select(
                Tag.id,
                Tag.title,
                func.count(Tag.id).label("games_count")
            )
            .join(GameTagAssociation, Tag.id == GameTagAssociation.tag_id)
            .filter(GameTagAssociation.game_id.in_(game_ids))
            .group_by(Tag.id, Tag.title)
        )

        if skip_tag_ids:
            stmt = stmt.filter(~Tag.id.in_(skip_tag_ids))

        stmt = (
            stmt.group_by(Tag.id, Tag.title)
            .order_by(desc(func.count(Tag.id)))
            .limit(limit)
        )

        result = await session.execute(stmt)

        tags = [ExtendedTag.from_orm(tag) for tag in result.all()]

        return tags


game_service = GameService()
